### PR TITLE
Fix #431 - management of `ALTER USER` when used with `IDENTIFIED WITH/VIA/BY` option

### LINKED ISSUE #431: `ALTER USER 'public'@'%' IDENTIFIED BY 'mypassword';` not well understood
Ref: https://github.com/phpmyadmin/phpmyadmin/issues/17654

Also: `ALTER USER 'public'@'%' IDENTIFIED WITH mysql_native_password BY '***'`

### PR DESCRIPTION
**Pull request by Tithugues**
This fix is linked to following issues:
* https://github.com/phpmyadmin/sql-parser/issues/431
* https://github.com/phpmyadmin/phpmyadmin/issues/17654

### CODE REVIEW

File: src/Components/AlterOperation.php [hunk 1]
```diff
@@ -159,8 +159,13 @@ class AlterOperation extends Component
             1,
             'var',
         ],
-        'BY' => [
+
```
Comment (niconoe-) [inline src/Components/AlterOperation.php:162]: I think this empy line is causing style issues.

File: src/Contexts/ContextMySql50700.php [hunk 2]
```diff
@@ -216,6 +216,7 @@ class ContextMySql50700 extends Context
         'FULLTEXT KEY' => 23, 'UNIQUE INDEX' => 23,
         'SPATIAL INDEX' => 23,
         'FULLTEXT INDEX' => 23,
+        'IDENTIFIED WITH' => 23,
```
Comment (niconoe-) [inline src/Contexts/ContextMySql50700.php:219]: You add this keyword only for MySQL 5.7 and MySQL 8.0, but there is also MariaDB contexts to upgrade IMO.

File: src/Contexts/ContextMySql80000.php [hunk 3]
```diff
@@ -217,6 +217,7 @@ class ContextMySql80000 extends Context
         'FULLTEXT KEY' => 23, 'UNIQUE INDEX' => 23,
         'SPATIAL INDEX' => 23,
         'FULLTEXT INDEX' => 23,
+        'IDENTIFIED WITH' => 23,
```
Comment (niconoe-) [inline src/Contexts/ContextMySql80000.php:220]: IIRC, those files are not intended to be manually updated but they are generated using `tools/contexts/*.txt files and calling the `tools/ContextGenerator.php` script.

In order to not loose your modifications here, you should update the txt files and run the generator.

File: src/Contexts/ContextMySql50700.php (hunk 2 above)
Comment (Tithugues) [reply src/Contexts/ContextMySql50700.php:219]: I don't think so. I checked the documentation and didn't find the same syntax.

Comment (niconoe-) [reply src/Contexts/ContextMySql50700.php:219]: https://mariadb.com/kb/en/alter-user/

Is defining this syntax for instance:
`ALTER USER <username> IDENTIFIED WITH <authentication_rule>`

I really do think it actually exists.

File: src/Contexts/ContextMySql80000.php (hunk 3 above)
Comment (Tithugues) [reply src/Contexts/ContextMySql80000.php:220]: Fixed!

File: src/Components/AlterOperation.php (hunk 1 above)
Comment (niconoe-) [reply src/Components/AlterOperation.php:162]: You're right, nevermind :+1:

**Comment by williamdes** [comment]
Can you rebase this PR for 5.8 without closing this PR or creating a new one?
I can do it on my end
All PRs should target 5.8 except if they are breaking changes for 6.0/master

**Comment by Tithugues** [comment]
> Can you rebase this PR for 5.8 without closing this PR or creating a new one? I can do it on my end All PRs should target 5.8 except if they are breaking changes for 6.0/master

Done! :pray:

**Review by niconoe- — APPROVED** [review]

File: tests/data/lexer/lexSymbolUser3.out [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+            "parseComment",
+            "parseOperator",
+            "parseBool",
+            "parseString",
+            "parseSymbol",
+            "parseKeyword",
+            "parseLabel",
+            "parseUnknown"
+        ],
+        "KEYWORD_NAME_INDICATORS": [
```
Comment (williamdes) [inline tests/data/lexer/lexSymbolUser3.out:18]: Seems like I forgot some of them in my custom json serialiser class in src tools

Comment (Tithugues) [reply tests/data/lexer/lexSymbolUser3.out:18]: Sorry, but I'm not sure… Do you expect anything from me here please?

Comment (williamdes) [reply tests/data/lexer/lexSymbolUser3.out:18]: No non, nothing
I just wanted to let you know that a file is doing some magic over the serialisation lib we use
and it's a todo for me to add the new properties I missed

**Review by williamdes — APPROVED** [review]
LGTM 💯 

@iifawzi what do you think about this one ?

File: src/Contexts/ContextMySql50700.php (hunk 2 above)
Comment (williamdes) [reply src/Contexts/ContextMySql50700.php:219]: On MariaDB 10.4
```sql
CREATE USER 'bob'@'localhost' 
  IDENTIFIED VIA mysql_native_password USING PASSWORD('vp8LAf4#wu2V&Wi*iJWC#3KPotsHzx3u') 
  OR unix_socket;
```
works
and
```sql
ALTER USER 'bob'@'localhost' IDENTIFIED VIA mysql_native_password USING PASSWORD('vp8LAf4#wu2V&Wi*iJWC#3KPotsHzx3u');
ALTER USER 'bob'@'localhost' IDENTIFIED WITH mysql_native_password USING PASSWORD('vp8LAf4#wu2V&Wi*iJWC#3KPotsHzx3u');
```
works too

File: tools/contexts/MariaDb100600.txt [hunk 5]
```diff
@@ -211,6 +211,7 @@ HOUR_MICROSECOND (R)
 HOUR_MINUTE (R)
 HOUR_SECOND (R)
 IDENTIFIED
+IDENTIFIED WITH (K)
```
Comment (iifawzi) [inline tools/contexts/MariaDb100600.txt:214]: I haven't seen (K) used before, what does it mean?

Comment (Tithugues) [reply tools/contexts/MariaDb100600.txt:214]: Honestly, I can't really explain.

You can find many `(K)` in this file: https://github.com/phpmyadmin/sql-parser/blob/master/tools/contexts/_common.txt

Additionally, you can find the flags here: https://github.com/phpmyadmin/sql-parser/blob/212ee22bef97747b3b2139cc2e13c9679a21d5db/src/Tools/ContextGenerator.php#L45 and they are used here: https://github.com/phpmyadmin/sql-parser/blob/212ee22bef97747b3b2139cc2e13c9679a21d5db/src/Tools/ContextGenerator.php#L147. And finally, the fact that `(K)` is associated to `16` seems to come from here: https://github.com/phpmyadmin/sql-parser/blob/212ee22bef97747b3b2139cc2e13c9679a21d5db/src/Token.php#L135

Comment (iifawzi) [reply tools/contexts/MariaDb100600.txt:214]: ohh.. didn't know it's used in common. 
awesome, thank you, lgtm then

**Review by iifawzi — APPROVED** [review]

**Comment by iifawzi** [comment]
Thank you for your continuous work on the package @Tithugues :')

File: src/Contexts/ContextMySql50700.php (hunk 2 above)
Comment (williamdes) [reply src/Contexts/ContextMySql50700.php:219]: @Tithugues can you add more tests for this two cases ?
Can this PR support them too ?

Comment (Tithugues) [reply src/Contexts/ContextMySql50700.php:219]: I'm on it. :+1:

Comment (Tithugues) [reply src/Contexts/ContextMySql50700.php:219]: `IDENTIFIED VIA` is now supported.

**Review by williamdes — APPROVED** [review]
Awesome work !

**Comment by williamdes** [comment]
Looking at 2b92a492aa885667736d340252d4442b0152cee4 
There is some stuff to add to remove the new errors

I now have some weird diffs in stage on full re-build
```sh
rm -r ./tests/data/**/*.out
./tools/run_generators.sh
```

Do you also have them ?

**Comment by williamdes** [comment]
Found it, probably caused by this PR.
Fixed by 3cf7606ad351abc687f6aed0c227536846cb5065

**Comment by williamdes** [comment]
Hi all,

I pushed documentation for context files: fbb763e947040f4849befe0189a9dbd1c83aee2f

**Pull request merged**